File: r/4.R
#!/usr/bin/env Rscript

library(digest)

mine <- function(input, test) {
    length <- nchar(test)
    getHash <- function(x) {
        substr(digest(paste(input, x, sep = ''),
                      algo = 'md5', serialize = FALSE),
               1, length)
    }

    i <- 1
    while (getHash(i) != test) {
        i <- i + 1
    }
    i
}

sideA <- function(input) {
    mine(input, '00000')
}

sideB <- function(input) {
    mine(input, '000000')
}

main <- function() {
    input <- 'ckczppom'
    cat(sideA(input), '\n')
    cat(sideB(input), '\n')
}

main()

File: js/4.js
#!/usr/bin/env node
'use strict'

const md5 = require('md5')
const fs = require('fs')

const parse = (str) => str.slice(0, -1)

const mine = (input, test) => {
  const hash = (str) => md5(str).slice(0, test.length)

  let i = 0
  for (; hash(input + i) != test; ++i)
    continue
  return i
}

const sideA = (str) => mine(str, '00000')

const sideB = (str) => mine(str, '000000')

const main = () => {
  const contents = fs.readFileSync('../4.txt', 'utf8')
  const input = parse(contents)
  console.log(sideA(input))
  console.log(sideB(input))
  return
}

main()
